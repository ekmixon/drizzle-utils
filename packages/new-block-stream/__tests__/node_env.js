/**
 * @jest-environment node
 */
// const Web3 = require("web3");
const { take, finalize, tap, toArray } = require("rxjs/operators");
const initTestChain = require("@drizzle-utils/test-chain");
const createNewBlock$ = require("../index");
const { blockMatcher } = require("./utils/propertyMatchers");

jest.setTimeout(20000);

describe("new-block-stream tests in node environment", () => {
  let provider;
  let web3;
  let accounts;
  let contractInstance;

  beforeAll(async () => {
    ({ provider, web3, accounts, contractInstance } = await initTestChain({
      contract: {
        dirname: __dirname,
        filename: "SimpleStorage.sol",
        contractName: "SimpleStorage",
      },
    }));
  });

  afterAll(async () => {
    provider.close();
  });

  test("createNewBlock$ function exists", () => {
    expect(createNewBlock$).toBeDefined();
  });

  test("createNewBlock$ throws errors when required options fields not found", () => {
    expect(() => createNewBlock$()).toThrow();

    expect(() => createNewBlock$({ pollingInterval: 200 })).toThrow(
      new Error("The options object with web3 is required."),
    );
  });

  test("fromPolling can track blocks", async done => {
    const newBlock$ = createNewBlock$({
      web3,
      pollingInterval: 200,
    });

    // tap observable to make sure it emitted a "0" and then a "5"
    newBlock$
      .pipe(
        take(2),
        toArray(),
        tap(vals => {
          expect(vals[0]).toMatchSnapshot(blockMatcher);
          expect(vals[1]).toMatchSnapshot(blockMatcher);
        }),
        finalize(() => {
          expect.assertions(2);
          done();
        }),
      )
      .subscribe();

    await contractInstance.methods.set(0).send({ from: accounts[0] });
    await contractInstance.methods.set(5).send({ from: accounts[0] });
  });

  // Remove test for subscribing to blocks, because we can't properly simulate
  // with Ganache, a provider that Web3 will allow us to subscribe to it.
  // test("fromSubscribe can track blocks", async done => {
  //   // Change constructor so we can test Websocket route
  //   // Note that ganache provider.constructor.name is Provider
  //   function WebsocketProvider() {}
  //   const web3Ws = new Web3(provider);
  //   web3Ws.currentProvider.constructor = WebsocketProvider;

  //   const newBlock$ = createNewBlock$({
  //     web3: web3Ws,
  //     pollingInterval: 200,
  //   });

  //   // tap observable to make sure it emitted a "0" and then a "5"
  //   newBlock$
  //     .pipe(
  //       take(2),
  //       toArray(),
  //       tap(vals => {
  //         expect(vals[0]).toMatchSnapshot(blockMatcher);
  //         expect(vals[1]).toMatchSnapshot(blockMatcher);
  //       }),
  //       finalize(() => {
  //         expect.assertions(2);
  //         done();
  //       }),
  //     )
  //     .subscribe();

  //   await contractInstance.methods.set(0).send({ from: accounts[0] });
  //   await contractInstance.methods.set(5).send({ from: accounts[0] });
  // });
});
