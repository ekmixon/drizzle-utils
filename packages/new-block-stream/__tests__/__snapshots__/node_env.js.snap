// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new-block-stream tests in node environment fromPolling can track blocks 1`] = `
Object {
  "difficulty": "0",
  "extraData": "0x",
  "gasLimit": 6721975,
  "gasUsed": Any<Number>,
  "hash": Anything,
  "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  "miner": "0x0000000000000000000000000000000000000000",
  "mixHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "nonce": "0x0000000000000000",
  "number": Any<Number>,
  "parentHash": Anything,
  "receiptsRoot": Any<String>,
  "sha3Uncles": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  "size": 1000,
  "stateRoot": Anything,
  "timestamp": Anything,
  "totalDifficulty": "0",
  "transactions": Array [
    Object {
      "blockHash": Anything,
    },
  ],
  "transactionsRoot": Any<String>,
  "uncles": Array [],
}
`;

exports[`new-block-stream tests in node environment fromPolling can track blocks 2`] = `
Object {
  "difficulty": "0",
  "extraData": "0x",
  "gasLimit": 6721975,
  "gasUsed": Any<Number>,
  "hash": Anything,
  "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  "miner": "0x0000000000000000000000000000000000000000",
  "mixHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "nonce": "0x0000000000000000",
  "number": Any<Number>,
  "parentHash": Anything,
  "receiptsRoot": Any<String>,
  "sha3Uncles": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  "size": 1000,
  "stateRoot": Anything,
  "timestamp": Anything,
  "totalDifficulty": "0",
  "transactions": Array [
    Object {
      "blockHash": Anything,
    },
  ],
  "transactionsRoot": Any<String>,
  "uncles": Array [],
}
`;
